File: src/test_prompt.py
#!/usr/bin/env python3
"""CLI утилита для тестирования генерации промпта из живой БД.

Использование:
    uv run -m src.test_prompt                      # последние 100 сообщений из дефолтного чата
    uv run -m src.test_prompt --limit 50           # последние 50 сообщений
    uv run -m src.test_prompt --from-id 220 --to-id 240  # конкретный диапазон
    uv run -m src.test_prompt --output prompt.txt  # вывод в файл
"""

import argparse
import asyncio
import sys

from .config import SYSTEM_PROMPT
from .db import db, get_messages
from .llm import generate_full_prompt

DEFAULT_CHAT_ID = -1001829561306
DEFAULT_LIMIT = 100


async def get_last_messages(chat_id: int, limit: int) -> list[dict]:
    """Получает последние N сообщений из чата."""
    return await db.fetch(
        f"""
        SELECT * FROM message 
        WHERE chat_id = <int> $chat_id
        ORDER BY message_id DESC
        LIMIT {int(limit)}
        """,
        {"chat_id": chat_id}
    )


async def main():
    parser = argparse.ArgumentParser(description="Генерирует промпт из живой БД")
    parser.add_argument("--chat-id", type=int, default=DEFAULT_CHAT_ID, help=f"ID чата (по умолчанию {DEFAULT_CHAT_ID})")
    parser.add_argument("--from-id", type=int, help="ID сообщения с которого начинать (exclusive)")
    parser.add_argument("--to-id", type=int, help="ID сообщения до которого (inclusive)")
    parser.add_argument("--limit", "-n", type=int, default=DEFAULT_LIMIT, help=f"Количество последних сообщений (по умолчанию {DEFAULT_LIMIT})")
    parser.add_argument("--output", "-o", type=str, help="Файл для вывода (по умолчанию stdout)")
    parser.add_argument("--full", action="store_true", help="Включить системный промпт")
    
    args = parser.parse_args()
    
    try:
        # Если указан диапазон - используем его, иначе берём последние N сообщений
        if args.from_id is not None and args.to_id is not None:
            messages = await get_messages(args.chat_id, args.from_id, args.to_id)
            range_info = f"from_id={args.from_id}, to_id={args.to_id}"
        else:
            messages = await get_last_messages(args.chat_id, args.limit)
            # Сортируем обратно в хронологическом порядке
            messages = list(reversed(messages))
            range_info = f"последние {args.limit}"
        
        if not messages:
            print(f"Сообщения не найдены для chat_id={args.chat_id}, {range_info}", file=sys.stderr)
            sys.exit(1)
        
        print(f"Найдено {len(messages)} сообщений (chat_id={args.chat_id})", file=sys.stderr)
        
        prompt = generate_full_prompt(messages)
        
        if args.full:
            prompt = SYSTEM_PROMPT + prompt
        
        if args.output:
            with open(args.output, "w") as f:
                f.write(prompt)
            print(f"Промпт записан в {args.output}", file=sys.stderr)
        else:
            print(prompt)
            
    finally:
        await db.close()


if __name__ == "__main__":
    asyncio.run(main())

File: src-ts/test-prompt.ts
/**
 * CLI утилита для тестирования генерации промпта из живой БД.
 *
 * Использование:
 *   bun run src-ts/test-prompt.ts                          # последние 100 сообщений из дефолтного чата
 *   bun run src-ts/test-prompt.ts --limit 50               # последние 50 сообщений
 *   bun run src-ts/test-prompt.ts --from-id 220 --to-id 240
 *   bun run src-ts/test-prompt.ts --output prompt.md
 */
import { parseArgs } from 'util';
import { Database } from 'bun:sqlite';
import { DB_PATH } from './config.ts';
import { generateFullPrompt, type MessageData } from './llm.ts';

const DEFAULT_CHAT_ID = -1001829561306;
const DEFAULT_LIMIT = 100;

const { values } = parseArgs({
  args: process.argv.slice(2),
  options: {
    'chat-id': { type: 'string', default: String(DEFAULT_CHAT_ID) },
    'from-id': { type: 'string' },
    'to-id':   { type: 'string' },
    'limit':   { type: 'string', short: 'n', default: String(DEFAULT_LIMIT) },
    'output':  { type: 'string', short: 'o' },
  },
});

const chatId = parseInt(values['chat-id']!);
const limit = parseInt(values['limit']!);
const fromId = values['from-id'] ? parseInt(values['from-id']) : undefined;
const toId = values['to-id'] ? parseInt(values['to-id']) : undefined;

const db = new Database(`${DB_PATH}/concisely.db`, { readonly: true });

let messages: MessageData[];
let rangeInfo: string;

if (fromId != null && toId != null) {
  messages = db.query<any, [number, number, number]>(
    'SELECT * FROM message WHERE chat_id = ? AND message_id > ? AND message_id <= ? ORDER BY message_id ASC',
  ).all(chatId, fromId, toId).map(parseRow);
  rangeInfo = `from_id=${fromId}, to_id=${toId}`;
} else {
  messages = db.query<any, [number, number]>(
    'SELECT * FROM message WHERE chat_id = ? ORDER BY message_id DESC LIMIT ?',
  ).all(chatId, limit).reverse().map(parseRow);
  rangeInfo = `последние ${limit}`;
}

if (!messages.length) {
  console.error(`Сообщения не найдены для chat_id=${chatId}, ${rangeInfo}`);
  process.exit(1);
}

console.error(`Найдено ${messages.length} сообщений (chat_id=${chatId})`);

const prompt = generateFullPrompt(messages);

if (values.output) {
  await Bun.write(values.output, prompt);
  console.error(`Промпт записан в ${values.output}`);
} else {
  console.log(prompt);
}

db.close();

function parseRow(r: any): MessageData {
  return {
    ...r,
    attachment: r.attachment ? JSON.parse(r.attachment) : null,
  };
}
